src/interactivity/base/HostSignalInputThread.cpp

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "HostSignalInputThread.hpp"
+#include "../inc/HostSignals.hpp"
+
+#include "../interactivity/inc/ServiceLocator.hpp"
+
+using namespace Microsoft::Console;
+using namespace Microsoft::Console::Interactivity;
+
+// Constructor Description:
+// - Creates the PTY Signal Input Thread.
+// Arguments:
+// - hPipe - a handle to the file representing the read end of the Host Signal pipe.
+HostSignalInputThread::HostSignalInputThread(_In_ wil::unique_hfile hPipe) :
+    _hFile{ std::move(hPipe) },
+    _hThread{},
+    _dwThreadId{ 0 },
+    _consoleConnected{ false }
+{
+    THROW_HR_IF(E_HANDLE, _hFile.get() == INVALID_HANDLE_VALUE);
+}
+
+HostSignalInputThread::~HostSignalInputThread()
+{
+    // Manually terminate our thread during unittesting. Otherwise, the test
+    //      will finish, but TAEF will not manually kill the test.
+#ifdef UNIT_TESTING
+    TerminateThread(_hThread.get(), 0);
+#endif
+}
+
+// Function Description:
+// - Static function used for initializing an instance's ThreadProc.
+// Arguments:
+// - lpParameter - A pointer to the HostSignalInputThread instance that should be called.
+// Return Value:
+// - The return value of the underlying instance's _InputThread
+DWORD WINAPI HostSignalInputThread::StaticThreadProc(_In_ LPVOID lpParameter)
+{
+    HostSignalInputThread* const pInstance = reinterpret_cast<HostSignalInputThread*>(lpParameter);
+    return pInstance->_InputThread();
+}
+
+// Method Description:
+// - Tell us that there's a client attached to the console, so we can actually
+//      do something with the messages we receive now. Before this is set, there
+//      is no guarantee that a client has attached, so most parts of the console
+//      (in and screen buffers) haven't yet been initialized.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HostSignalInputThread::ConnectConsole() noexcept
+{
+    _consoleConnected = true;
+}
+
+// Method Description:
+// - The ThreadProc for the Host Signal Input Thread.
+// Return Value:
+// - S_OK if the thread runs to completion.
+// - Otherwise it may cause an application termination another route and never return.
+[[nodiscard]] HRESULT HostSignalInputThread::_InputThread()
+{
+    unsigned short signalId;
+    while (_GetData(&signalId, sizeof(signalId)))
+    {
+        switch (signalId)
+        {
+        case HOST_SIGNAL_NOTIFY_APP:
+        {
+            HOST_SIGNAL_NOTIFY_APP_DATA msg = { 0 };
+            if (_GetData(&msg, sizeof(msg)))
+            {
+                if (msg.cbSize >= sizeof(msg))
+                {
+                    _AdvanceReader(msg.cbSize - sizeof(msg));
+                }

```

Comment:
this (and the exception) seems like shared code between the three cases; is there an opportunity for templatization here? It can easily figure out the size of the struct and advance the correct distance.